rfcs/src/rfcs/0001-mir-linker.md

```diff
@@ -0,0 +1,244 @@
+ by pruning the inputs of symtab2gb and goto-instrument.
+  - Linking against the standard library goto-models takes more than 5 GB which is not feasible with the current
+  GitHub workers.
+- In a potential assessment mode, only analyze code that is reachable from all public items in the target crate.
+
+One downside is that our release bundle will be bigger.
+We are going to include compilation artifacts for the `std` and `kani` libraries.
+This will negatively impact the time taken to set up Kani
+(triggered by either the first time a user invokes `kani | cargo-kani` , or explicit invoke the subcommand `setup`).
+
+## User Experience
+
+Once this RFC has been stabilized users shall use Kani in the same manner as they have been today.
+Until then, we wil add an unstable option `--mir-linker` to enable the cross-crate reachability analysis
+and the generation of the `goto-program` model only when compiling the target crate.
+
+Kani setup will likely take longer and more disk space as mentioned in the section above.
+This change will not be guarded by `--mir-linker` option above.
+
+## Detailed Design
+
+In a nutshell, we will no longer generate a `goto-program` model for every crate we compile.
+Instead, we will generate the `MIR` for every crate, and we will generate only one `goto-program` model.
+This model will only include items reachable from the target crate's harnesses.
+
+The current system flow for a crate verification is the following (Kani here represents either `kani | cargo-kani`
+executable):
+
+1. Kani compiles the user crate as well as all its dependencies.
+   For every crate compiled, `kani-compiler` will generate a `goto-program` model.
+   This model includes everything reachable from the crate's public functions.
+2. After that, Kani links all models together by invoking `goto-cc`.
+   This step will also link against Kani's `C` library.
+3. For every harness, Kani prunes the linked model to only include items reachable from the given harness.
+4. Finally, Kani instruments and verify each harness model via `goto-program` and `cbmc` calls.
+
+After this RFC, the system flow would be slightly different:
+
+1. Kani compiles the user crate dependencies up to the `MIR` translation.
+   I.e., for every crate compiled, `kani-compiler` will generate an artifact that includes the `MIR` representation
+  of all items in the crate.
+2. Kani will generate the `goto-program` only while compiling the target user crate.
+  It will generate one `goto-program` model that includes all items reachable from any harness in the target crate.
+3. `goto-cc` will still be invoked to link the generated model against Kani's `C` library.
+4. Steps #3 and #4 above will be performed without any change.
+
+This feature will require three main changes to Kani which are detailed in the sub-sections below.
+
+### Kani's Sysroot
+
+Kani currently uses `rustup` sysroot to gather information from the standard library constructs.
```

Comment:
I'm a bit unclear on this section. What standard library constructs are looked up in `sysroot`? Does `sysroot` have MIR lying around?

Reply:
Yes. The rustup sysroot does have MIR lying around. They are inside the *rlib files shipped with the toolchain (see more details here: https://rustc-dev-guide.rust-lang.org/backend/libs-and-metadata.html#metadata).

Reply:
@zhassan-aws what kind of information would you like me to add here to make this more clear?

Reply:
I think it's fine as is.